motion_predict/src/motion_predict.cpp

```diff
@@ -0,0 +1,147 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+#include "motion_predict.h"
+
+namespace Motion{
+
+cav_msgs::PredictedState MotionPredict::predictState(const geometry_msgs::Pose& pose, const geometry_msgs::Twist& twist,const double delta_t)
+{
+  Eigen::VectorXd x(4); // State Vector
+
+  x(0)=pose.position.x; // Position X
+  x(1)=pose.position.y; // Position Y
+  x(2)=twist.linear.x; // Linear Velocity X
+  x(3)=twist.linear.y; // Linear Velocity Y
+  
+  Eigen::MatrixXd F=Eigen::MatrixXd::Identity(x.size(), x.size()); // Generate identity matrix for state transition matrix
+
+  F(0,2)=delta_t; 
+  F(1,3)=delta_t;
+  
+  x = F * x; // Predict 
+
+  cav_msgs::PredictedState pobj;
+
+  pobj.predicted_position.position.x=x(0); // Predicted Position X
+  pobj.predicted_position.position.y=x(1); // Predicted Position Y
+  pobj.predicted_position.position.z=pose.position.z; // Predicted Position Z
+
+  pobj.predicted_velocity.linear.x=x(2); // Predicted Linear Velocity X
+  pobj.predicted_velocity.linear.y=x(3); // Predicted Linear Velocity Y
+  pobj.predicted_velocity.linear.z=twist.linear.z; // Predicted Linear Velocity Z
+
+  return pobj;
+} 
+
+
+// Forward predict an external object
+cav_msgs::PredictedState MotionPredict::externalPredict(const cav_msgs::ExternalObject &obj,const double delta_t,const double ax,const double ay,const float process_noise_max)
+{
+
+  cav_msgs::PredictedState pobj = predictState(obj.pose.pose, obj.velocity.twist,delta_t);
+
+  Eigen::MatrixXd F=Eigen::MatrixXd::Identity(4,4); // Generate identity matrix for state transition matrix
+
+  F(0,2)=delta_t; 
+  F(1,3)=delta_t;
+
+  Eigen::MatrixXd P(4,4); // State Covariance Matrix
+  P.fill(0.0);  // Filling whole matrix with zero
+
+  // Covariance extraction
+  P(0,0)=obj.pose.covariance[0]; // X
+  P(1,1)=obj.pose.covariance[7]; // Y
+  P(2,2)=obj.velocity.covariance[0]; // Vx
+  P(3,3)=obj.velocity.covariance[7]; // Vy
+
+  Eigen::MatrixXd Q(4,4); // Process Noise Matrix
+
+  double delta_t2 = delta_t * delta_t; //t^2
+  double delta_t3 = delta_t2 * delta_t;//t^3 
+  double delta_t4 = delta_t3 * delta_t;//t^4
+
+  Q << (delta_t4 / 4 * ax), 0,( delta_t3 / 2 * ax), 0, 0, (delta_t4 / 4 * ay), 0,( delta_t3 / 2 * ay), (delta_t3 / 2 * ax), 0,(delta_t2*ax), 0 , 0 , (delta_t3 / 2 * ay), 0 , (delta_t2*ay); // Process Noise Matrix
+  
+  Eigen::MatrixXd Ft = F.transpose(); // Transpose of State Transition Function
+    
+  P = F * P * Ft + Q; //State Covariance Matrix
+
+  float position_process_noise_avg=(P(0,0)+P(1,1))/2; // Position process noise average
+
+  pobj.predicted_position_confidence=Mapping(position_process_noise_avg,process_noise_max); // Position process noise confidence
+
+  float velocity_process_noise_avg=(P(2,2)+P(3,3))/2; // Position velocity process noise average
```

Comment:
Might as well make this and the other variable doubles. Lets keep things as accurate as possible. It's over optimizing at this point to make them floats. If there is a performance issue we will go back and optimize then.